util/deephash/deephash.go

```diff
@@ -303,17 +308,38 @@ func makeSliceHasher(t reflect.Type) typeHasherFunc {
 
 	var once sync.Once
 	var hashElem typeHasherFunc
+	var isRecursive bool
 	init := func() {
 		hashElem = lookupTypeHasher(t.Elem())
+		isRecursive = typeIsRecursive(t)
 	}
 
 	return func(h *hasher, p pointer) {
 		pa := p.sliceArray()
+		if pa.isNil() {
+			h.HashUint8(0) // indicates nil
+			return
+		}
 		once.Do(init)
+		h.HashUint8(1) // indicates visiting slice
 		n := p.sliceLen()
 		h.HashUint64(uint64(n))
 		for i := 0; i < n; i++ {
 			pe := pa.arrayIndex(i, nb)
+			if isRecursive {
+				func() {
+					if idx, ok := h.visitStack.seen(pe.p); ok {
+						h.HashUint8(2) // indicates cycle
+						h.HashUint64(uint64(idx))
+						return
+					}
+					h.HashUint8(1) // indicates visiting slice element
+					h.visitStack.push(pe.p)
+					defer h.visitStack.pop(pe.p)
+					hashElem(h, pe)
+				}()
```

Comment:
these immediately-called-funcs are kinda gross style.

just for the defer?

Reply:
Yea. I realized looking at the code earlier that it was buggy without it. Did you have an alternative suggestion?

Reply:
Using method? Not using a defer?

Reply:
I realized that the recursive case is identical to normal hashing of elements, but with additional logic before.
Implemented it as a wrapper around the original `hashElem`.